hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -2079,6 +2060,21 @@ int SaraNcpClient::enterDataMode() {
     return r;
 }
 
+int SaraNcpClient::dataModeError(int error) {
+    if (ncpId() == PLATFORM_NCP_SARA_R410 && error == SYSTEM_ERROR_PPP_NO_CARRIER_IN_NETWORK_PHASE) {
+        CHECK_TRUE(connectionState() == NcpConnectionState::CONNECTED, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_FALSE(cgattWorkaroundApplied_, SYSTEM_ERROR_INVALID_STATE);
```

Comment:
So the behavior will be when we get `NO CARRIER` during PPP, this handler will be called and `cgattWorkaroundApplied_` is checked. We will try the `AT+CGATT` workaround and if it succeeds, the flag is cleared when the NCP connection state transitions to `NcpConnectionState::CONNECTED`

If we try the workaround and it still does not connect, the `AT+CGATT` workaround will be skipped until it can 'naturally' connect on its own (and then clearing the `cgattWorkaroundApplied_` flag)?

Reply:
`cgattWorkaroundApplied_` only clears when we get registered. `connectionState()` is pretty much CEREG state, so the for all intents and purposes it gets cleared when we get deregistered/re-register, power cycle etc.

Reply:
PPP connection state is separate and NCP client doesn't really care that much about it, except for facilitating when dialing.